src/source-pcap-folder.c

```diff
@@ -0,0 +1,943 @@
+    if (pcap_g.checksum_mode == CHECKSUM_VALIDATION_DISABLE) {
+        p->flags |= PKT_IGNORE_CHECKSUM;
+    } else if (pcap_g.checksum_mode == CHECKSUM_VALIDATION_AUTO) {
+        if (ChecksumAutoModeCheck(ptv->pkts, p->pcap_cnt,
+                                  SC_ATOMIC_GET(pcap_g.invalid_checksums))) {
+            pcap_g.checksum_mode = CHECKSUM_VALIDATION_DISABLE;
+            p->flags |= PKT_IGNORE_CHECKSUM;
+        }
+    }
+
+    PACKET_PROFILING_TMM_END(p, TMM_RECEIVEPCAPFOLDER);
+
+    if (TmThreadsSlotProcessPkt(ptv->tv, ptv->slot, p) != TM_ECODE_OK) {
+        pcap_breakloop(pcap_g.pcap_handle);
+        ptv->cb_result = TM_ECODE_FAILED;
+    }
+
+    SCReturn;
+}
+
+int PcapFolderGetModifiedTime(char const * file, time_t * out)
+{
+    struct stat buf;
+    int ret;
+
+    if (file == NULL)
+        return -1;
+
+    SCLogInfo("PcapFolderGetModifiedTime \"%s\"", file);
+
+    if ((ret = stat(file, &buf)) != 0)
+        return ret;
+
+    *out = buf.st_mtime;
+
+    return ret;
+}
+
+int PcapFolderSortByStatTime(const void * vleft, const void * vright)
+{
+    if (vleft == vright)
+        return 0;
+
+    time_t leftTime = 0, rightTime = 0;
+
+    int leftRet, rightRet;
+
+    leftRet = PcapFolderGetModifiedTime(*(const char **)vleft, &leftTime);
+    rightRet = PcapFolderGetModifiedTime(*(const char **)vright, &rightTime);
+
+    if (leftRet == 0 && rightRet == 0)
+    {
+        return (leftTime == rightTime) ? 0 : ((leftTime < rightTime) ? -1 : 1);
+    }
+    else
+    {
+        return (leftRet == rightRet) ? 0 : ((leftRet < rightRet) ? -1 : 1);
+    }
+}
+
+void PcapFolderFreeMemBuffer(MemBuffer * buffer)
+{
+    if (buffer == NULL)
+        return;
+    
+    size_t offset = 0;
+
+    while (offset < MEMBUFFER_OFFSET(buffer))
+    {
+        SCFree(*((char**)(MEMBUFFER_BUFFER(buffer) + offset)));
+        offset += sizeof(char*);
+    }
+    MemBufferFree(buffer);
+}
+
+/**
+ *  \brief Main PCAP file reading Loop function
+ */
+TmEcode ReceivePcapFolderLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    int packet_q_len = 64;
+    PcapFolderThreadVars *ptv = (PcapFolderThreadVars *)data;
+    int r;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+    ptv->cb_result = TM_ECODE_OK;
+
+    time_t newerThan = 0;
+
+    while (1) {
+        if (suricata_ctl_flags & SURICATA_STOP) {
+            SCReturnInt(TM_ECODE_OK);
+        }
+
+        /* make sure we have at least one packet in the packet pool, to prevent
+         * us from alloc'ing packets at line rate */
+        PacketPoolWait();
+
+        time_t const fileDelay = 30;
+        time_t olderThan = time(NULL);
+        olderThan -= fileDelay;
+        time_t firstNewFileTime = 0;
+        time_t lastProcessed = 0;
+
+        DIR * directory = NULL;
+
+        directory = opendir(pcap_g.dirpath);
+        if (directory == NULL)
+        {
+            switch (errno)
+            {
+                case EACCES:
```

Comment:
bad indent in this block. Also it might be good to move in a util func as this function is overly long